packages/rest-crud/src/crud-rest.controller.ts

```diff
@@ -254,8 +241,43 @@ export function defineCrudRestController<
     }
   }
 
-  // See https://github.com/microsoft/TypeScript/issues/14607
-  return CrudRestControllerImpl;
+  if (options && options.idRequired) {
```

Comment:
At runtime, we already know whether the model requires the id to be provided by the user or whether it will be generated by the database. Let's use that information please and don't ask the user to supply it to us again.

Relevant code from LB3:
- https://github.com/strongloop/loopback-swagger/blob/fe5cfd7a11fb4cb0c9804db510d461a3d2ad53a6/lib/specgen/model-helper.js#L46-L60
- https://github.com/strongloop/loopback/blob/aad97c20360de236af6afcb4e364f42d071664b0/lib/model.js#L794-L799
- https://github.com/strongloop/loopback-datasource-juggler/blob/de4718d5b81d29ac23597172d8ee591617924fe0/lib/model.js#L841-L844
- https://github.com/strongloop/loopback-datasource-juggler/blob/de4718d5b81d29ac23597172d8ee591617924fe0/lib/model-builder.js#L343-L362

Please note that we need to support more update-only properties, for example the Cloudant connector uses property `_rev` that is always generated by the database.

I see this is topic is more complex than I initially thought :(

If you prefer, then I am ok to leave the changes in `packages/rest-crud` out of scope of this pull request, so that it can be landed sooner. Just open a follow-up issue and add it to the epic #2036 _From model definition to REST API with no custom repository/controller classes_ please.

Reply:
Sounds good. Please open two stories - one to update `rest-crud`, another to improve the CLI tool.